runtime/common/src/claims.rs

```diff
@@ -570,13 +623,98 @@ mod tests {
 				InvalidTransaction::Custom(ValidityError::SignerHasNoClaim.into()).into(),
 			);
 		});
 	}
 }
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking {
+	use super::*;
+	use secp_utils::*;
+	use system::RawOrigin;
+	use frame_benchmarking::{benchmarks, account};
+	use sp_runtime::DispatchResult;
+	use sp_runtime::traits::ValidateUnsigned;
+	use crate::claims::Call;
+
+	const SEED: u32 = 0;
+
+	const MAX_CLAIMS: u32 = 10_000;
+	const VALUE: u32 = 1_000_000;
+
+	fn create_claim<T: Trait>(input: u32) -> DispatchResult {
+		let secret_key = secp256k1::SecretKey::parse(&keccak_256(&input.encode())).unwrap();
+		let eth_address = eth(&secret_key);
+		let vesting = Some((100_000.into(), 1_000.into(), 100.into()));
+		super::Module::<T>::mint_claim(RawOrigin::Root.into(), eth_address, VALUE.into(), vesting)?;
+		Ok(())
+	}
+
+	benchmarks! {
+		_ {
+			// Create claims in storage.
+			let c in 0 .. MAX_CLAIMS => create_claim::<T>(c)?;
+		}
+
+		// Benchmark `claim` for different users.
+		claim {
+			let u in 0 .. 1000;
+			let secret_key = secp256k1::SecretKey::parse(&keccak_256(&u.encode())).unwrap();
+			let eth_address = eth(&secret_key);
+			let account: T::AccountId = account("user", u, SEED);
+			let vesting = Some((100_000.into(), 1_000.into(), 100.into()));
+			let signature = sig::<T>(&secret_key, &account.encode());
+			super::Module::<T>::mint_claim(RawOrigin::Root.into(), eth_address, VALUE.into(), vesting)?;
+		}: _(RawOrigin::None, account, signature)
+
+		// Benchmark `mint_claim` when there already exists `c` claims in storage.
+		mint_claim {
+			let c in ...;
+			let account = account("user", c, SEED);
+			let vesting = Some((100_000.into(), 1_000.into(), 100.into()));
+		}: _(RawOrigin::Root, account, VALUE.into(), vesting)
+
+		// Benchmark the time it takes to execute `validate_unsigned` 
+		validate_unsigned {
+			let c in ...;
+			// Crate signature
+			let secret_key = secp256k1::SecretKey::parse(&keccak_256(&c.encode())).unwrap();
+			let account: T::AccountId = account("user", c, SEED);
+			let signature = sig::<T>(&secret_key, &account.encode());
+			let call = Call::<T>::claim(account, signature);
+		}: {
+			super::Module::<T>::validate_unsigned(&call)?
+		}
+
+		// Benchmark the time it takes to do `repeat` number of keccak256 hashes
+		keccak256 {
+			let i in 0 .. 10_000;
+			let bytes = (i).encode();
+		}: {
+			for index in 0 .. i {
+				let _hash = keccak_256(&bytes);
+			}
+		}
+
+		// Benchmark the time it takes to do `repeat` number of `eth_recover` 
+		eth_recover {
+			let i in 0 .. 1_000;
+			// Crate signature
+			let secret_key = secp256k1::SecretKey::parse(&keccak_256(&i.encode())).unwrap();
+			let account: T::AccountId = account("user", i, SEED);
+			let signature = sig::<T>(&secret_key, &account.encode());
+			let data = account.using_encoded(to_ascii_hex);
+		}: {
+			for _ in 0 .. i {
+				let _maybe_signer = super::Module::<T>::eth_recover(&signature, &data);
```

Comment:
I wonder if the compiler optimize some part of it?

maybe it could be interesting to use https://docs.rs/criterion/0.3.1/criterion/fn.black_box.html , though I'm very not sure

If benchmarks are run with wasm runtime then this won't get optimized for the interesting parts so it is ok I guess

Reply:
Yes, it is always run with the wasm runtime. I can if needed use the variable so that it for sure does not get optimized away, but I have not seen results which imply it is skipped.